text/3120-preview-features.md

```diff
@@ -0,0 +1,520 @@
+shortage of users willing to try out the feature and provide experience
+reports with using it. That, in turn, provides more feedback for honing
+its design and implementation in the early phases, and provides more
+evidence for maturity and usefulness in the later phases. That evidence
+is ultimately what is needed to build the confidence needed to stabilize
+a feature, so such features have a robust path to stabilization. Less
+highly desired features get less attention, and may spend much longer
+before the get stabilized if at all. This is, arguably, the process
+working as designed -- features that users want get prioritized.
+
+There is, however, a class of features that suffers under this process
+-- those that impact a large number of _users_ without impacting a large
+number of Rust _developers_. Consider a feature like Cargo's
+[`-Zstrip`], which strips debug symbols from binaries and thus
+significantly reduces their size. It's likely not a feature a large
+number of individual Rust developers care about, but once it's made
+available, developers who disseminate programs written in Rust can much
+more easily distribute smaller binaries to all of their users, who may
+in turn have a better experience from the faster downloads. Another
+example is [`-Zpatch-in-config`], which allow non-Cargo build systems to
+inject `[patch]` sections into Rust crates when they're built. Few
+individual developers need this feature, but the developers of those
+build systems may be able to use that feature to significantly improve
+the experience of working with first-party Rust dependencies under their
+build tool, which may in turn improve the user experience of large
+number of Rust developers that in turn rely on that build system.
+
+The reason this class of features is underserved by the current process
+is two-fold. First, since few developers need them _directly_, the
+perceived need for the feature is less than that of a high-profile
+feature like const generics, even if the actual _impact_ of the feature
+may be comparable. Second, and more importantly, users who are in the
+position to make use of such high-impact changes are often hesitant to
+abandon the stability guarantees of Rust's stable channel. While Rust's
+nightly releases are _generally_ quite stable, and flags like
+`-Zallow-features` exist to limit the unstableness of the nightly
+compiler, at such large scale even the slight risk from switching to
+nightly is often unacceptable. For this reason, the developers who could
+make use of these high-impact features may be unable to actually use
+those features until they're stable (see [rationale and
+alternatives][rationale-and-alternatives] for alternatives). But since
+stabilization requires evidence of impact and maturity, these features
+can get caught in a Catch-22 situation where evidence won't be provided
+until they're stabilized, and they can't be stabilized without evidence.
```

Comment:
My biggest meta-question here is whether we've sufficiently tried *asking* for more evidence on these things.

For example, the "enabled in one place in a build system" feature of cargo pipelining, where as far as I can tell, asking for people to try it out and give feedback in <https://internals.rust-lang.org/t/evaluating-pipelined-rustc-compilation/10199> was highly successful without any special technical mechanism.

How much of the value here could we get from something that wouldn't need an RFC at all, such as an occasional blog post asking people to try out certain features on a particular known-to-be-reasonable nightly?

Reply:
It's a good point! I think a combination of a recommended, well-chosen nightly plus an explicit reminder to set `-Zallow-features` gets us quite far. But at the same time, I feel like "recommended, well-chosen nightly" is really just a suboptimal version of `env RUSTC_BOOSTRAP=1 cargo +stable` (or `+beta` maybe rather).

To be clear, I'm not opposed to recommending `RUSTC_BOOSTRAP` for these kinds of use-cases. My biggest grip with it is that it opts into _everything_, and then users have to know to opt out with `-Zallow-features`. _And_ they have to make a (fairly tricky) choice for each feature whether or not to allow it. My thinking with preview features was basically that rather than relegate that choosing to users, it's made by those who are best equipped to decide the stability (and fragility) of a feature.

Reply:
> And they have to make a (fairly tricky) choice for each feature whether or not to allow it.

I don't really see that.  If I use the same example above, I don't get the impression that people started turning on other stuff as part of it.  I would expect that a post saying "hey, try out this cool thing by running `RUSTC_FLAGS=-Zstrip cargo +nightly build`" would generally have people doing just that, not also suddenly adding some other random features like `#[feature(link_llvm_intrinsics)]` or something.

(Especially in a big corporate setting, where they're all the more likely to have strict CI and code reviews and static analysis and be using their own builds instead of shipped ones because of toolchain attack concerns and ...)